Added check to prevent infinite looping with multiapps

DanielYankura opened the pull request:
<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->
## Reason
<!--Why do you need this feature or what is the enhancement?-->
- Having an infinite multiapp loop can waste resources and be hard to debug
## Design
<!--A concise description (design) of the enhancement.-->
- Added param max_multiapp_level which can be modified by user
- Error is thrown if multiapp level exceeds the set limit
## Impact
<!--Will the enhancement change existing APIs or add something new?-->
- Can aid in debugging input files

<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->
closes #16803
also closes https://github.com/idaholab/moose/issues/24371

DanielYankura committed:
Added check to prevent infinite looping with multiapps

- Added param max_multiapp_level which can be modified by user
- Error is thrown if multiapp level exceeds the set limit

closes #16803

GiudGiud commented on framework/src/multiapps/MultiApp.C:
```diff
@@ -226,6 +226,11 @@ MultiApp::validParams()
+      "max_multiapp_level",
+      10,
+      "Integer set by user that will stop the simulation if the multiapp level "
+      "exceeds it. Usefull for preventing infinite loops with multiapp simulations");
```
```suggestion
      "exceeds it. Useful for preventing infinite loops with multiapp simulations");
```

GiudGiud replied:
let's add it to the Advanced param group please

GiudGiud commented on test/tests/multiapps/full_solve_multiapp/tests:
```diff
@@ -79,5 +79,13 @@
+      expect_err = "Maximum multiapp level has been reached. This is likely caused by an infinite loop "
+               "in your multiapp system. If additional multiapp levels are needed, "
+               "max_multiapp_level can be specified in the MuliApps block."
+      detail = "Throw error is max_multiapp_level is reached"
```
```suggestion
      detail = "throw an error is the user-selected maximum level of nesting of applications is reached"
```

GiudGiud reviewed (commented):
nice. did not know about this level API

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2653197/), step Docs: sync website on a10a215 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/29757/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2653213/), step Generate coverage on a10a215 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>951f23</th>
      <th colspan=4>#29757 a10a21</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.25%</td>
      <td>85.25%</td>
      <td>+0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>108121</td>
      <td>108125</td>
      <td>+4</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18714</td>
      <td>18714</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/29757/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/29757/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/29757/coverage/xfem)

</details>



This comment will be updated on new commits.

DanielYankura committed:
small changes for PR

-corrected spelling and added more to test detail
-added max_multiapp_level to advanced param group

GiudGiud commented:
Test failures are unrelated
